swaylock/main.c

```diff
@@ -417,56 +455,160 @@ int main(int argc, char **argv) {
+				cairo_stroke(window->cairo);
+			} break;
+			case AUTH_STATE_VALIDATING: {
+				cairo_set_source_rgba(window->cairo, 0, 114.0 / 255, 255.0 / 255, 0.75);
+				cairo_fill_preserve(window->cairo);
+				cairo_set_source_rgb(window->cairo, 51.0 / 255, 0, 250.0 / 255);
+				cairo_stroke(window->cairo);
+			} break;
+			case AUTH_STATE_INVALID: {
+				cairo_set_source_rgba(window->cairo, 250.0 / 255, 0, 0, 0.75);
+				cairo_fill_preserve(window->cairo);
+				cairo_set_source_rgb(window->cairo, 125.0 / 255, 51.0 / 255, 0);
+				cairo_stroke(window->cairo);
+			} break;
+			default: break;
+			}
+
+			// Draw a message
+			char *text = NULL;
+			cairo_set_source_rgb(window->cairo, 0, 0, 0);
+			cairo_set_font_size(window->cairo, ARC_RADIUS/3.0f);
+			switch (render_data->auth_state) {
+			case AUTH_STATE_VALIDATING:
+				text = "verifying";
+				break;
+			case AUTH_STATE_INVALID:
+				text = "wrong";
+				break;
+			default: break;
+			}
+
+			if (text) {
+				cairo_text_extents_t extents;
+				double x, y;
+
+				cairo_text_extents(window->cairo, text, &extents);
+				x = window->width/2 - ((extents.width/2) + extents.x_bearing);
+				y = window->height/2 - ((extents.height/2) + extents.y_bearing);
+
+				cairo_move_to(window->cairo, x, y);
+				cairo_show_text(window->cairo, text);
+				cairo_close_path(window->cairo);
+				cairo_new_sub_path(window->cairo);
+			}
+
+			// Typing indicator: Highlight random part on keypress
+			if (render_data->auth_state == AUTH_STATE_INPUT || render_data->auth_state == AUTH_STATE_BACKSPACE) {
+				double highlight_start = (rand() % (int)(2 * M_PI * 100)) / 100.0;
+				cairo_arc(window->cairo, window->width/2, window->height/2, ARC_RADIUS, highlight_start, highlight_start + TYPE_INDICATOR_RANGE);
+				if (render_data->auth_state == AUTH_STATE_INPUT) {
+					cairo_set_source_rgb(window->cairo, 51.0 / 255, 219.0 / 255, 0);
+				} else {
+					cairo_set_source_rgb(window->cairo, 219.0 / 255, 51.0 / 255, 0);
+				}
+				cairo_stroke(window->cairo);
+
+				// Draw borders
+				cairo_set_source_rgb(window->cairo, 0, 0, 0);
+				cairo_arc(window->cairo, window->width/2, window->height/2, ARC_RADIUS, highlight_start, highlight_start + TYPE_INDICATOR_BORDER_THICKNESS);
+				cairo_stroke(window->cairo);
+
+				cairo_arc(window->cairo, window->width/2, window->height/2, ARC_RADIUS, highlight_start + TYPE_INDICATOR_RANGE, (highlight_start + TYPE_INDICATOR_RANGE) + TYPE_INDICATOR_BORDER_THICKNESS);
+				cairo_stroke(window->cairo);
+			}
+
+			// Draw inner + outer border of the circle
+			cairo_set_source_rgb(window->cairo, 0, 0, 0);
+			cairo_set_line_width(window->cairo, 2.0);
+			cairo_arc(window->cairo, window->width/2, window->height/2, ARC_RADIUS - ARC_THICKNESS/2, 0, 2*M_PI);
+			cairo_stroke(window->cairo);
+			cairo_arc(window->cairo, window->width/2, window->height/2, ARC_RADIUS + ARC_THICKNESS/2, 0, 2*M_PI);
+			cairo_stroke(window->cairo);
+		}
```

Comment:
Just a minor comment, I guess most of the cairo stuff could be done once outside the loop, instead of doing the same drawing for every output.